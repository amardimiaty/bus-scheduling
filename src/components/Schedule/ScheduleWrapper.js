// Packages
import styled from 'styled-components'


const ScheduleWrapper = styled.div`
  font-family: 'Lato';
  padding-top: 48px;
  width: 725px;
  margin: 0 auto;
  border-bottom: 1px solid #BDBDBD;
  .Schedule__operating-hrs {
    position: relative;
    left: 140px;
    font-size: 12px;
    margin-bottom: 12px;
    span {
      display: inline-block;
      width: 25px;
    }
    span:not(:first-child) {
      // Need to figure out better way to derive this #
      margin-left: 33px;
    }
  }
`

export default ScheduleWrapper