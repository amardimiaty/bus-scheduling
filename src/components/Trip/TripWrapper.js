// Packages
import styled from 'styled-components'

const TripWrapper = styled.div`
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 4px;
  height: 30px;
  border: 1px solid #9E9E9E;
  width: ${props => props.endTime - props.startTime}px;
  left: ${props => props.startTime + 140}px;
  background-color: ${props => props.isSelected ? '#EFFAFF' : 'white'};
  cursor: pointer;
`

export default TripWrapper