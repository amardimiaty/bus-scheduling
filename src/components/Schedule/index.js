// Packages
import React, { useState, useEffect, createContext } from 'react'

// Data
import BusSchedule from '../../data/schedule.json'
import OperatingHrs from '../../data/operatingHours.json'

// Styled Components
import ScheduleWrapper from './ScheduleWrapper'

// Components
import Bus from '../Bus'

const ScheduleContext = createContext()
const ScheduleProvider = ScheduleContext.Provider

const Schedule = () => {
  const schedule = BusSchedule
  const operatingHours = OperatingHrs
  const [buses, updateBuses] = useState([])
  const [selectedTrip, updateSelected] = useState({})
  const [isNewBusShowing, updateIsNewBusShowing] = useState(false)

  /* eslint-disable */
  useEffect(() => {
    // Initial seeding.
    let busesToAdd = []
    schedule.map((trip, idx) => busesToAdd.push({number: idx + 1,  trips: [trip] }))
    updateBuses(busesToAdd)
  }, [])

  useEffect(() => {
    // If the selected trip changes and it is not empty show new bus.
    if(selectedTrip.id) {
      if(!isNewBusShowing) {
        const busesWithNew = [...buses, { number: 'New Bus', trips: [] }]
        updateBuses(busesWithNew)
        updateIsNewBusShowing(true)
      }
    } else {
      updateIsNewBusShowing(false)
    }
  }, [selectedTrip])
/* eslint-enable */

  return (
    <ScheduleProvider value={{
      schedule,
      selectedTrip, 
      updateSelected,
      buses,
      updateBuses,
    }}>
      <ScheduleWrapper>
        <div className="Schedule__operating-hrs">
          {
            operatingHours.map(h => (<span key={h}>{h}</span>))
          }
        </div>
        {
          buses.map((bus, idx) => (
            <Bus key={`bus-${idx + 1}`} trips={bus.trips} number={bus.number}/>
          ))
        }
      </ScheduleWrapper>
    </ScheduleProvider>
  )
}

export default Schedule
export { ScheduleContext, ScheduleProvider }