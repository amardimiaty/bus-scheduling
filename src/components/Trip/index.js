// Packages
import React, { useContext } from 'react'

// Context
import { ScheduleContext } from '../Schedule'

// Styled Components
import TripWrapper from './TripWrapper'

const Trip = ({ trip, busNumber }) => {
  const {selectedTrip, updateSelected} = useContext(ScheduleContext)
  return (
    <TripWrapper
      isSelected={trip.id === selectedTrip.id}
      endTime={trip.endTime} 
      startTime={trip.startTime}
      onClick={() => updateSelected({ 
        id:trip.id, 
        currentBus: busNumber, 
        endTime: trip.endTime, 
        startTime: trip.startTime 
      })}>
        <span className="Trip__number">{trip.id}</span>
    </TripWrapper>
  )
}

export default Trip