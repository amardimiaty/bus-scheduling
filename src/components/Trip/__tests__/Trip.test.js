// Packages
import React from 'react'
import wait from 'waait'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'

// Context
import { ScheduleProvider } from '../../Schedule'

// Components
import Trip from '../index'

const fakeTrip = {
 'id': 1, 
 'startTime': 30, 
 'endTime': 150
}

const fakeContext = {
  selectedTrip: {},
  update: () => {},
}

describe('<Trip />', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <ScheduleProvider value={fakeContext}>
        <Trip busNumber={1} trip={fakeTrip} />
      </ScheduleProvider>
    )
    await wait()
    wrapper.update()
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('displays bus number', async () => {
    const wrapper = mount(
      <ScheduleProvider value={fakeContext}>
        <Trip busNumber={1} trip={fakeTrip} />
      </ScheduleProvider>
    )
    await wait()
    wrapper.update()
    const tripNumber = wrapper.find('.Trip__number')
    expect(tripNumber.text()).toContain('1');
  })
});