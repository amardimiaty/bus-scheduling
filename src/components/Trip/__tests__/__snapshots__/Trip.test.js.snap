// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Trip /> renders and matches snapshot 1`] = `
<Trip
  busNumber={1}
  trip={
    Object {
      "endTime": 150,
      "id": 1,
      "startTime": 30,
    }
  }
>
  <styled.div
    endTime={150}
    isSelected={false}
    onClick={[Function]}
    startTime={30}
  >
    <StyledComponent
      endTime={150}
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": false,
            "lastClassName": "jHRzPn",
            "rules": Array [
              "
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 4px;
  height: 30px;
  border: 1px solid #9E9E9E;
  width: ",
              [Function],
              "px;
  left: ",
              [Function],
              "px;
  background-color: ",
              [Function],
              ";
  cursor: pointer;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      isSelected={false}
      onClick={[Function]}
      startTime={30}
    >
      <div
        className="sc-htpNat jHRzPn"
        onClick={[Function]}
      >
        <span
          className="Trip__number"
        >
          1
        </span>
      </div>
    </StyledComponent>
  </styled.div>
</Trip>
`;
